### PR TITLE
MetaClientCache API's

### PR DESCRIPTION
**Pull request by Marcosrico**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

https://github.com/apache/helix/issues/2237

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR includes all API's, datastructures, and config modifications to prepare lattice for cache adoption.

### Tests

- [x] The following tests are written for this issue:

`TestZkMetaClientCache.java`

Tested that the cachclient is created and works for write requests and fails for read requests (as it should read for cache and the implementation will be built in a future PR).

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+     * @param cacheData Indicates whether the data should be cached.
+     * @param cacheChildren Indicates whether the children of the nodes should be cached.
+     * @param lazyCaching Indicates whether lazy loading of the cache is enabled.
+     */
+    public ZkMetaClientCache(ZkMetaClientConfig config, String key, Boolean cacheData,
+                             Boolean cacheChildren, Boolean lazyCaching) {
+        super(config);
+        _rootEntry = key;
+        _lazyLoading = lazyCaching;
+        _cacheData = cacheData;
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java:68]: Should we enforce to be true?

File: meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    public void closeCache() {
+        throw new MetaClientException("Not implemented yet.");
+    }
+
+    @Override
+    public void rebuildCache() {
+        throw new MetaClientException("Not implemented yet.");
+    }
+
+    @Override
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java:92]: Add API doc. (Also all following APIs)

File: meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientCacheInterface.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * Close / end the cache.
+     */
+    void closeCache();
+
+    /**
+     * Rebuild the entries that exist in the current cache. Will log an error and return null
+     * if the cache is close / hasn't started yet.
+     */
+    void rebuildCache();
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientCacheInterface.java:55]: I feel like this API is a bit confusing. Any existing use case now? I am open to discussion.

File: meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientCacheInterface.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+         * When a change happens in the underlying metadata storage, the
+         * cache must be updated accordingly. In this state, the cache is
+         * being rebuilt based on the changes that the listener picks up.
+         */
+        PROCESSING_CHANGES,
+        /**
+         * The cache is up-to-date with the key from the underlying
+         * metadata storage and no changes are being processed.
+         */
+        UP_TO_DATE
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientCacheInterface.java:37]: Where is the state being consumed?

**Comment by Marcosrico** [comment]
After an offline discussion we agreed on a couple comments for this design:
1. There is no need for the apis (start, close, and rebuild cache) in the MetaClientCacheInterface, as those methods are essentially done in the original MetaClientCache that such extends. Thus, we can safely remove the need for a cache interface.
2. Instead, we will add a cache field to the metaclientconfig and the builder that will create an instance of the zkmetaclientcache if enabled.
3. There is also no need to expose the state of the cache to the user. The api contract is pre-defined and so whether the state of the cache is up-to-date or not is not needed.

File: meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/factory/ZkMetaClientFactory.java [hunk 5]
```diff
@@ -36,4 +38,17 @@ public MetaClientInterface getMetaClient(MetaClientConfig config) {
     }
     throw new IllegalArgumentException("Invalid MetaClientConfig type.");
   }
+
+  @Override
+  public MetaClientCacheInterface getMetaClientCache(MetaClientConfig config, String key, Boolean cacheData,
+                                                     Boolean cacheChildren, Boolean lazyCaching) {
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/factory/ZkMetaClientFactory.java:44]: Can we group the later 4 vars into a cache config?

File: meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    public List<T> get(List<String> keys) {
+        throw new MetaClientException("Not implemented yet.");
+    }
+
+    @Override
+    public List<Stat> exists(List<String> keys) {
+        throw new MetaClientException("Not implemented yet.");
+    }
+
+    public ZkClient getCacheClient() {
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java:101]: Why we need this API?

File: meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/factory/ZkMetaClientFactory.java (hunk 5 above)
Comment (Marcosrico) [reply meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/factory/ZkMetaClientFactory.java:44]: Are you saying to have a cache config and a zk cache config too?

File: meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java (hunk 6 above)
Comment (Marcosrico) [reply meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java:101]: Not sure if there's a use case right now but it is just a getter method (like in the original zk client there is a getZkClient() method)

Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java:101]: If there is no usage, we dont need to expose this. If we need to expose this for test, we can add "visibleForTestOnly"

File: meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/factory/ZkMetaClientFactory.java (hunk 5 above)
Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/factory/ZkMetaClientFactory.java:44]: I was thinking having a cachConfig that has the folling 4 vars.
String key, Boolean cacheData, Boolean cacheChildren, Boolean lazyCaching

File: meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java (hunk 6 above)
Comment (Marcosrico) [reply meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientCache.java:101]: Sounds good, will remove unnecessary api's for now.

File: meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientCacheConfig.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+public class MetaClientCacheConfig {
+    private String _rootEntry;
+    private boolean _cacheData;
+    private boolean _cacheChildren;
+    private boolean _lazyCaching;
+
+    public MetaClientCacheConfig(String rootEntry, boolean cacheData, boolean cacheChildren, boolean lazyCaching) {
+        _rootEntry = rootEntry;
+        _cacheData = cacheData;
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientCacheConfig.java:32]: Add default values for all fields except _rootEntry

**Review by xyuanlu — APPROVED** [review]
Generally looks good. Please address comments.

**Comment by Marcosrico** [comment]
PR approved by @xyuanlu
Commit message:
MetaClientCache Part 1 - API's, configs, and builders

File: meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientFactory.java [hunk 8]
```diff
@@ -49,4 +50,20 @@ public MetaClientInterface getMetaClient(MetaClientConfig config) {
...
+      throw new IllegalArgumentException("MetaClientConfig cannot be null.");
+    }
+    if (MetaClientConfig.StoreType.ZOOKEEPER.equals(config.getStoreType())) {
+      ZkMetaClientConfig zkMetaClientConfig = new ZkMetaClientConfig.ZkMetaClientConfigBuilder().
+              setConnectionAddress(config.getConnectionAddress())
+              .setMetaClientReconnectPolicy(config.getMetaClientReconnectPolicy())
+              .setConnectionInitTimeoutInMillis(config.getConnectionInitTimeoutInMillis())
+              .setSessionTimeoutInMillis(config.getSessionTimeoutInMillis())
+              .build();
+      return new ZkMetaClientFactory().getMetaClientCache(zkMetaClientConfig, cacheConfig);
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientFactory.java:65]: Forgot to mention it in prev comment.  line 60-65 can be merged with line 44-49. Please address it in following PR.

**Pull request merged**